### ISSUE TITLE
Terminate an action

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#1249]
## Problem

Dagger has the ability to run commands in containers (`engine.#Exec`). These commands can be long-running (for example a test database). But there is no reliable way to terminate this command later. This effectively makes it impossible to execute long-running commands without hanging DAG execution.

A typical example (from real-world use case #1227):
* Start a test database (`engine.#Exec`)
* Execute tests against test database (`engine.Exec`)
* After the tests aere completed, terminate the test database (*this is the missing feature*)

## Solution

Here’s a suggestion to get us started:

```
package engine

// Terminate a task
#Terminate: {
  // The ID of the task to terminate
  id: #TaskID
}

// Each task produces a unique ID which can be safely referenced
// NOTE: not necessarily the same as our internal IDs
#Exec: {
  // Dagger task ID.
  target: #TaskID
  …
}
```

Example:

```
sleep: engine.#Exec & {
  args: [“sleep”, “100”]
}

terminate: engine.#Terminate & {
  target: sleep.id
}
```

NOTE: in this example, `sleep` and `terminate` will run in parallel, causing `sleep` to be terminated immediately. For a more useful example, we need a way to run `terminate` after other commands. This requires [explicit dependencies between tasks](https://github.com/dagger/dagger/issues/1248).

**Comment by sipsma** [comment]
Just some initial thoughts having looked into this a bit.

> When a task is terminated, what is the expectation after that occurs? I'm guessing it's ok if we throw away filesystem state (no cache), or have any state we'd like to keep around in a cachedir mount?

Given that the ability to terminate the vertex at arbitrary points introduces significant non-determinism, I think we definitely don't want to cache the results of these types of vertices in the same way normal `Exec` results are cached. That being said, we can still support getting the filesystem state after the execution finishes, we just need to make sure that the filesystem state isn't going to be re-used by other solves.

Shared cache mounts would, like you suggested, work, especially for cases like this database one where all the modified state the user is interested in is most likely centralized into one directory (i.e. /var/db/). The only slight downside here is that cache mounts don't fit use cases where the user wants to capture all state created to arbitrary places in the rootfs; a given cache mount can only be placed at a single directory and will block out any filesystem on the underlying rootfs present at that dir.

If we wanted to support getting any filesystem state created by the terminated vertex, there's probably a path to doing so. Namely, [the `Solve` method of Buildkit actually allows you to get the results of failed vertices](https://github.com/moby/buildkit/pull/1732), which can be used temporarily (for the life of the current gateway client instance I believe).
* The problem is we most likely want to use `NewContainer` rather than `Solve` for these long-running terminable vertices, as that interface allows you to send signals (i.e. `SIGTERM`) to the container explicitly, and `NewContainer` doesn't have that same support for temporary, non-cached results. 
* However, looking at the implementation, I suspect it would be fairly easy to re-use the same machinery from `Solve` to add that support to `NewContainer`. The only harder part might be if we want to export these states, which (unless I'm missing a simpler path) would require [this issue](https://github.com/moby/buildkit/issues/1914) to be addressed.

Overall, my opinion right now is
1. It probably makes sense to use `NewContainer` for this as it cleanly supports explicit stopping of the vertex
2. Initially, we can support users who want to get state created by their terminated vertex by having them use shared cache mounts
3. In the longer term, more generic solutions for getting the filesystem results of these types of vertices is possible too with some upstream modifications to BuildKit.

**Comment by sipsma** [comment]
Talked with @jlongtine about the Exec case some more yesterday, specifically the cases similar to the test DB example where you want an exec to run asynchronously while some other vertices use it and then terminate the exec once it's not needed anymore.

Given the differences in caching behavior between these async execs and a normal exec (discussed in my previous comment), I'm inclined to think it makes sense to have these async execs be a different cue type than normal execs. It seems like the original proposal of this issue was for a more generic ability to terminate any type of action, but I can't currently see a way to do that while always maintaining cache coherency; I think the ability to terminate may need to be added selectively to individual task types that can support it and that will not cache their non-deterministic results. Let me know if anyone thinks differently.

To start, for simplicity I'm going to try implementing the start and stop of an async exec as separate tasks, i.e. there will be a `StartAsyncExec` vertex, then a set of test vertices that depend on that, then a `StopAsyncExec` that depends on the completion of the tests. A rough version of this should give me a better idea of the issues at play (and how to split this issue up into reasonably sized changesets rather than a big blob of code).

For the final product, I'd be interested to see if we could abstract the separate Start/Stop steps from users and instead just have them define a single vertex (like `db`) that is then transformed into separate start/stop steps by our cue evaluator underneath the hood based on the dependencies it can see. Don't know enough about cue yet to say if that's actually possible though (cc @jlongtine).

**Comment by sipsma** [comment]
Have the most primitive-but-working possible implementation here: https://github.com/sipsma/dagger/commit/6810f7783b821b32814f1656437249c432192b30

The cue I'm using to test is:
```
package main

import (
	"dagger.io/dagger"
	"dagger.io/dagger/core"
)

dagger.#Plan & {

	actions: {
		test: {
			image: core.#Pull & {
				source: "alpine:3.15"
			}

			asyncExec: core.#AsyncExec & {
				input: image.output
				args: [
					"sh", "-c", "sleep 30",
				]
			}

			sleep: core.#Exec & {
				input: image.output
				args: [
					"sleep", "3",
				]
				always: true
			}

			stop: core.#StopAsyncExec & {
				execID: asyncExec.execID
				// TODO: extreme hack, enables explicit dep
				wait: sleep.exit
			}
		}
	}
}
```

Basically, the async exec does `sleep 30`, then a normal exec sleeps for only 3 seconds and once it exits the `StopAsyncExec` task runs (had to use a hack to add a dependency, pending #1248). When you run this the `test` action exits after 3 seconds (and you can see via debug logs that the async exec is actually killed).

There's more error cases to handle, temp hacks (like a global map) to clean up, etc. But it does seem like the basic idea of using `NewContainer` and `AsyncExec` is at least technically sound.

**Comment by sipsma** [comment]
> That's awesome 😃 We definitely need https://github.com/dagger/dagger/issues/1248.

Thanks! And I agree we need that, I knew it was going to be an issue but didn't realize how fundamental it would be until I wrote it out. I'm gonna get this implementation to the point where it can be a draft PR and then consider looking into #1248 to unblock this.

**Comment by shykes** [comment]
This will make Dagger much more useful for running integration tests 👍👍👍

**Comment by sipsma** [comment]
Resolved in #2214, more follow-ups tracked here: https://github.com/dagger/dagger/issues/2398
